core/src/main/java/org/infinispan/statetransfer/StateTransferInterceptor.java

```diff
@@ -186,20 +187,37 @@ public void init(StateTransferManager stateTransferManager) {
          while (ce instanceof RemoteException) {
             ce = ce.getCause();
          }
-         if (!(ce instanceof OutdatedTopologyException) && !(ce instanceof SuspectException))
-            throw throwable;
-
-         // We increment the topology id so that updateTopologyIdAndWaitForTransactionData waits for the next topology.
-         // Without this, we could retry the command too fast and we could get the OutdatedTopologyException again.
-         if (trace)
-            log.tracef("Retrying command because of topology change, current topology is %d: %s",
+         if (ce instanceof OutdatedTopologyException) {
+            if (trace)
+               log.tracef("Retrying command because of topology change, current topology is %d: %s",
                   currentTopologyId(), rCommand);
-         FlagAffectedCommand flagAffectedCommand = (FlagAffectedCommand) rCommand;
-         int newTopologyId = Math.max(currentTopologyId(), flagAffectedCommand.getTopologyId() + 1);
-         flagAffectedCommand.setTopologyId(newTopologyId);
-         waitForTopology(newTopologyId);
-
-         return visitReadCommand(rCtx, flagAffectedCommand, consistentHashUpdater);
+            // We increment the topology id so that updateTopologyIdAndWaitForTransactionData waits for the next topology.
+            // Without this, we could retry the command too fast and we could get the OutdatedTopologyException again.
+            FlagAffectedCommand flagAffectedCommand = (FlagAffectedCommand) rCommand;
+            int newTopologyId = Math.max(currentTopologyId(), flagAffectedCommand.getTopologyId() + 1);
+            flagAffectedCommand.setTopologyId(newTopologyId);
+            CompletableFuture<Void> future = stateTransferLock.topologyFuture(newTopologyId);
+            if (future != null) {
+               return future.thenCompose(ignored -> {
+                  try {
+                     return visitReadCommand(rCtx, flagAffectedCommand, consistentHashUpdater);
+                  } catch (Throwable t) {
+                     throw new CacheException(t);
+                  }
+               });
+            }
+         } else if (ce instanceof SuspectException) {
+            if (trace)
+               log.tracef("Retrying command because of suspected node, current topology is %d: %s",
+                  currentTopologyId(), rCommand);
+            // Upon SuspectException we don't wait for newer topology, as it is possible that current topology
+            // is actual but the view still contains a node that's about to leave; a broadcast to all nodes
+            // then can end with suspect exception, but we won't get any new topology. An example of this situation
+            // is when a node sends leave - topology can be installed before the new view.
```

Comment:
The opposite can be a problem as well: in a distributed cache, we don't broadcast write commands but instead send them to the owners in the current topology. If a node crashes and we get a `SuspectException`, we don't want to go in a busy loop, retrying the command and failing with a new `SuspectException` until we receive the new topology. (Not least, because using lots of CPU retrying will delay the processing of the new topology.)

We work around the problem you mention here by always using `ResponseMode.SYNCHRONOUS_IGNORE_LEAVERS` when we broadcast a cache command. That way, we can never get a `SuspectException` after we've already updated our topology.